### ISSUE TITLE
Shouldn't WithArguments(object) filter out null properties before sending in the request?

### ISSUE THREAD

**Issue by Ninglin** [Pathoschild/FluentHttpClient#73]
If we pass an object with null properties on it, those will be sent with the request with the value string.Empty. Should this behavior be like this? I think null properties should be ignored or the behavior should be configurable. Something like `IgnoreNullProperties = true` when instancing the Client.

**Comment by Pathoschild** [comment]
How about something like this to keep the number of constructors down?

```cs
var client = new FluentClient(baseUrl, new FluentClientOptions
{
   IgnoreNullArguments = true,
   IgnoreHttpErrors = false,
   Proxy = ...
});
```

**Comment by Ninglin** [comment]
That would be a good approach. What would be the most desirable default for this? I would say true is the more correct approach. What are your thoughts about this?

**Comment by Pathoschild** [comment]
That's fine with me. @Jericho any preference on whether to ignore `null` arguments by default?

**Comment by Jericho** [comment]
My opinion is to ignore them by default.

One quick comment: since this setting is so specific to `WithArguments(object)`, doesn't it make sense to add a (optional) parameter to this method rather than a property when instanciating the client? Something like this:

```csharp
IRequest WithArguments(object arguments, bool ignoreNullArguments = true);
```

**Comment by Pathoschild** [comment]
I see it as a set-and-forget option, not something you'd want to set on each `WithArguments` call. An `options` constructor argument would be a good place for finetuning options like this without cluttering the rest of the API.

**Comment by Ninglin** [comment]
I would say that the better option is to have this at the the client instantiation. This is a behaviour for all requests within a given `FluentClient` so it makes sense to configure this within the `FluentClientOptions`.

**Comment by Jericho** [comment]
I don't use this method so the final decision will not affect me but I could see a situation where someone could want the null arguments of one object to be ignored _and also_ the null arguments of another object to be included in the same request

``` csharp
var response = await client
    .GetAsync("qwerty.php")
    .WithArguments(new { a = "aaa", b = null }, true)
    .WithArguments(new { c = "ccc", d = null }, false)
    .WithArguments(new { e = "eee", f = null })   // No value specified, default value applies
    .AsMessage();
```
 (by the way, the fact that a default value is set on the argument in my sample means that you wouldn't have to worry about it on each call but it gives you the option to override the default value on any call).

**Comment by Pathoschild** [comment]
@Jericho You could do that by explicitly setting an empty string, which won't be ignored:
```cs
await client
    .GetAsync("qwerty.php")
    .WithArguments(new { a = null,  b = "" }); // qwerty.php?b=
```

**Comment by Jericho** [comment]
@Pathoschild what about cases where the data type is not `string`?

Of the top of my head, something like this:
```csharp
await client
    .GetAsync("qwerty.php")
    .WithArguments(new { a = null,  b = (int?)null });
```

**Comment by Pathoschild** [comment]
In cases where you want to control individual arguments you could do this:
```cs
await client
    .GetAsync("qwerty.php")
    .WithArguments(new
    {
        a = null,
        b = value?.ToString() ?? ""
    }); // qwerty.php?b=
```

**Comment by Jericho** [comment]
We keep using an anonymous object built on the fly which gives us flexibility but not all cases are like that. Consider the following:

```csharp
var myObject = new MyModelClass()
{
    a = "aaa",
    b = null    // Let's assume data type is int?
};
await client
    .GetAsync("qwerty.php")
    .WithArguments(myObject);
```

You could say: create a dynamic object on the fly like this:
```csharp
await client
    .GetAsync("qwerty.php")
    .WithArguments(new
    {
        a = myObject.a,
        b = myObject.b?.ToString() ?? ""
    });
```

but this becomes tedious quickly when the model class has several properties.

Anyway, like I said, I'm not using this method so feel free to take my opinion with a grain of salt but I just think we should accommodate as many scenarios as possible.

**Comment by Pathoschild** [comment]
That's a fair point, but I'm not sure how common the use case will be. If users rarely need to control null handling per-argument with a model, then creating an anonymous object on the fly is fine. I suggest releasing without a `WithArguments` option, and we can add it in a future version if there's demand for it. Any objections?

**Comment by Jericho** [comment]
No objection

**Comment by Ninglin** [comment]
Doesn't the `WithArgument` method serve the purpose of forcing something to be null. For instance, if someone would want to force a parameter to be null on the same client wouldn't they be able to call `WithArgument("parameterName", null)`?

**Comment by Jericho** [comment]
@Ninglin I was just going to ask: for consistency sake, don't we want `WithArgument` to respect the `IgnoreNullArguments` settings?
 
Otherwise we would have a global `IgnoreNullArguments` setting that is enforced in `WithArguments` but not in `WithArgument`. Personally, I would be confused by that.

**Comment by Ninglin** [comment]
@Jericho I had the same question when asking that

**Comment by Pathoschild** [comment]
If you set `IgnoreNullArguments = true`, then `WithArgument("parameterName", null)` will have no effect. Since multiple argument values are allowed, this won't remove a previous argument:
```cs
await client
    .GetAsync("example")
    .WithArgument("arg", 1)
    .WithArgument("arg", 2)
    .WithArgument("arg", null); // example?arg=1&arg=2
```

**Comment by Jericho** [comment]
There's a third method that deals with arguments where I think we should consistently enforce the IgnoreNullArguments setting:

```csharp
IRequest WithArguments<TKey, TValue>(IEnumerable<KeyValuePair<TKey, TValue>> arguments);
```

**Comment by Pathoschild** [comment]
Yep, null arguments should be ignored regardless of how they're passed in.

**Comment by Jericho** [comment]
To summarize: all three `WithArgument(s)` methods will consistently enforce the `IgnoreNullArguments` setting. Right?

**Comment by Pathoschild** [comment]
Yep.

**Comment by Ninglin** [comment]
Sounds good.

**Comment by Jericho** [comment]
Submitted a PR which allows to specify `IgnoreNullArguments` when instantiating the client and also allows to override this value for each request if desired.

**Comment by Pathoschild** [comment]
Merged into `develop` for the upcoming 3.2 release. Null arguments will be ignored by default, and you can override that via `client.SetOptions` or `request.WithOptions`. Thanks!

**Comment by Pathoschild** [comment]
Per discussion in #76, we'll keep the previous behaviour by default to avoid breaking changes. You can set `client.SetOptions(ignoreNullArguments: true)` to use the new behavior in 3.2. The default behavior will change in the future 4.0 release.
